Added tutorial and example

royderks opened the pull request.

jwm0 commented on saleor-react-ts/TUTORIAL.md:
````diff
@@ -0,0 +1,478 @@
+
+We will use the [Create React App](https://github.com/facebook/create-react-app) to get quickly started. To use Create React App you need to globally install it on your computer, if you haven't already, with:
+
+```bash
+yarn global add create-react-app
+```
````
This doesn’t (and shouldn’t) need to be installed, `npx` fetches the most recent version temporarily before executing the command. You can remove that part

royderks replied:
Ah nice

jwm0 commented on saleor-react-ts/TUTORIAL.md:
````diff
@@ -0,0 +1,478 @@
+```tsx
+// src/Products.tsx
+import { useQuery } from '@apollo/client';
+import { loader } from 'graphql.macro';
+
+const getLatestProducts = loader('src/getLatestProducts.graphql');
````
It seems like using the `gql` inside js/ts files is the more common/preferred approach. I suggest renaming the file `src/productQueries.ts` and exporting it as:
```
// productQueries.ts
export const LATEST_PRODUCTS = gql`<query goes here>`
```

royderks replied:
Yes I often prefer this, but `graphql-codegen` works with `.graphql` files..

jwm0 replied:
`graphql-codegen` can also work with `ts` files if you're using the `gql` macro. See here: https://graphql-code-generator.com/docs/getting-started/documents-field#available-formats

jwm0 commented on saleor-react-ts/TUTORIAL.md:
```diff
@@ -0,0 +1,478 @@
+
+    return (
+      <div>
+        {latestProducts?.length &&
```
For an empty array this will result in a `0` printed out
```suggestion
        {latestProducts?.length > 0 &&
```

jwm0 commented on saleor-react-ts/TUTORIAL.md:
```diff
@@ -0,0 +1,478 @@
+
+This query should be used as input for an Apollo `useQuery` Hook, that returns three variables: `loading`, `data`, and `error`. These variables are pretty self-explanatory and can be used to define what you want to return from your React application.
+
+The `useQuery` Hook can be imported from `@apollo/client`, just as the `gql` function to parse the GraphQL query language in your application. These can be imported in a new file called `Products.js` in the `src` directory:
```
`gql` is mentioned here but not used in the example
`Products.js` => `Products.tsx`

jwm0 commented on saleor-react-ts/TUTORIAL.md:
```diff
@@ -0,0 +1,478 @@
+
+In this file, the query `getLatestProducts` is passed to the `useQuery` Hook and used to retrieve the last 5 products from the GraphQL API. The data that's returned is being destructed and rendered in the `Products` component. We haven't added any type definitions to this file, which we definitely should as this is a TypeScript application.
+
+The `useQuery` Hook accepts a generic for the return type of the data that it's fetching which, as we're using GraphQL, has the same shape as the GraphQL query. You can use the GraphQL Playground to check what the shape of the output is, in case you're in doubt. Afterwards, add the new type `LatestProduct` in the `Products` component and add it to the Hook:
```
We should definitely mention the ability to generate the TS files straight from the server schema, it's one of major sellers for TS + GraphQL stack 🙌 . I see you're using it later on in the tutorial but I think it's worth a quick mention here.

jwm0 commented on saleor-react-ts/TUTORIAL.md:
```diff
@@ -0,0 +1,478 @@
+const getLatestProducts = loader('src/getLatestProducts.graphql');
+
+function Products() {
+  const { loading, error, data } = useQuery<LatestProducts>(getLatestProducts);
```
`LatestProducts` type probably shouldn't be here until we mention it later in the tutorial, right?

jwm0 commented on saleor-react-ts/TUTORIAL.md:
````diff
@@ -0,0 +1,478 @@
+  // ...
+```
+
+To render this in the browser we need to alter the `src/App.js` file and replace its contents with the following:
````
all files should have `.ts(x)` file format

jwm0 commented on saleor-react-ts/src/Products.tsx:
```diff
@@ -0,0 +1,41 @@
+};
+
+function Products({ keyword = 'juice' }: ProductsProps) {
+  const { loading, error, data } = useGetLatestProductsQuery({
```
query is not typed

royderks replied:
Actually it is, the "magic" is in `src/generated/graphql.tsx` as a strong typed hook is exported :)

krzysztofwolski commented on saleor-react-ts/src/Products.tsx:
```diff
@@ -0,0 +1,41 @@
+                <h3>{name}</h3>
+                <p>{description}</p>
+                <ul>
+                  <li>{category?.name}</li>
```
List for a single element seems weird.

royderks commented:
> <img alt="obraz" width="957" src="https://user-images.githubusercontent.com/5188636/110485355-d4281a80-80eb-11eb-9fe1-b77fd9bf383a.png">
> 
> Result page would use some UI improvemeents. Just a bit of css and font would make it much better

Yes might be a nice addition, I believed earlier on we dropped this to focus on creating the TS types. Would you prefer adding Bootrap or Tailwind?

royderks commented:
> > <img alt="obraz" width="957" src="https://user-images.githubusercontent.com/5188636/110485355-d4281a80-80eb-11eb-9fe1-b77fd9bf383a.png">
> > Result page would use some UI improvemeents. Just a bit of css and font would make it much better
> 
> Yes might be a nice addition, I believed earlier on we dropped this to focus on creating the TS types. Would you prefer adding Bootrap or Tailwind? @krzysztofwolski

Added some basic bootstrap styling
<img width="1359" alt="Screenshot 2021-04-14 at 17 55 25" src="https://user-images.githubusercontent.com/10717410/114740940-99e51500-9d4a-11eb-9da7-4c3012a3e3ee.png">

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/saleor/saleor-examples?pullRequest=1) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/saleor/saleor-examples?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/saleor/saleor-examples?pullRequest=1) it.</sub>
